pkg/ccl/importccl/read_import_mysql.go

```diff
@@ -143,5 +150,197 @@ func mysqlToCockroach(
+	if in == nil {
+		return nil, errors.Errorf("could not read definition for table %q (possible unsupoprted type?)", name)
+	}
+
+	time := hlc.Timestamp{WallTime: evalCtx.GetStmtTimestamp().UnixNano()}
+	priv := sqlbase.NewDefaultPrivilegeDescriptor()
+	desc := sql.InitTableDescriptor(id, defaultCSVParentID, name, time, priv)
+
+	colNames := make(map[string]string)
+	for _, raw := range in.Columns {
+		def, err := mysqlColToCockroach(raw.Name.String(), raw.Type)
+		if err != nil {
+			return nil, err
+		}
+		col, _, _, err := sqlbase.MakeColumnDefDescs(def, &tree.SemaContext{}, evalCtx)
+		if err != nil {
+			return nil, err
+		}
+		desc.AddColumn(*col)
+		colNames[raw.Name.String()] = col.Name
+	}
+
+	for _, raw := range in.Indexes {
+		var elems tree.IndexElemList
+		for _, col := range raw.Columns {
+			elems = append(elems, tree.IndexElem{Column: tree.Name(colNames[col.Column.String()])})
+		}
+		idx := sqlbase.IndexDescriptor{Name: raw.Info.Name.String(), Unique: raw.Info.Unique}
+		if err := idx.FillColumns(elems); err != nil {
+			return nil, err
+		}
+		if err := desc.AddIndex(idx, raw.Info.Primary); err != nil {
+			return nil, err
+		}
+	}
+
+	// AllocateIDs mutates its receiver. `return desc, desc.AllocateIDs()`
+	// happens to work in gc, but does not work in gccgo.
+	//
+	// See https://github.com/golang/go/issues/23188.
+	err := desc.AllocateIDs()
+	return &desc, err
+}
+
+func mysqlColToCockroach(name string, col mysql.ColumnType) (*tree.ColumnTableDef, error) {
+	def := &tree.ColumnTableDef{Name: tree.Name(name)}
+
+	var length, scale int
+
+	if col.Length != nil {
+		num, err := strconv.Atoi(string(col.Length.Val))
+		if err != nil {
+			return nil, errors.Wrapf(err, "could not parse length for column %q", name)
+		}
+		length = num
+	}
+
+	if col.Scale != nil {
+		num, err := strconv.Atoi(string(col.Scale.Val))
+		if err != nil {
+			return nil, errors.Wrapf(err, "could not parse scale for column %q", name)
+		}
+		scale = num
+	}
+
+	switch col.SQLType() {
+	case mysqltypes.Char:
+		def.Type = strOpt(coltypes.Char, length)
+	case mysqltypes.VarChar:
+		def.Type = strOpt(coltypes.VarChar, length)
+	case mysqltypes.Text:
+		def.Type = strOpt(coltypes.Text, length)
+	case mysqltypes.Blob:
+		def.Type = coltypes.Bytes
+	case mysqltypes.VarBinary:
+		def.Type = coltypes.Bytes
+	case mysqltypes.Binary:
+		def.Type = coltypes.Bytes
+
+	case mysqltypes.Int8:
+		def.Type = coltypes.SmallInt
+	case mysqltypes.Uint8:
+		def.Type = coltypes.SmallInt
+	case mysqltypes.Int16:
+		def.Type = coltypes.SmallInt
+	case mysqltypes.Uint16:
+		def.Type = coltypes.SmallInt
+	case mysqltypes.Int24:
+		def.Type = coltypes.Int
+	case mysqltypes.Uint24:
+		def.Type = coltypes.Int
+	case mysqltypes.Int32:
+		def.Type = coltypes.Int
+	case mysqltypes.Uint32:
+		def.Type = coltypes.Int
+	case mysqltypes.Int64:
+		def.Type = coltypes.BigInt
+	case mysqltypes.Uint64:
+		def.Type = coltypes.BigInt
+
+	case mysqltypes.Float32:
+		def.Type = coltypes.Float4
```

Comment:
Is there a difference here? I thought floats are all 64 bit in cockroach regardless. It seems silly to keep around a mysql-ism only in the visible type when the underlying type is really just a float.

Reply:
We have a bunch of code around preserving the different visible types for the same underlying type, so I assumed we wanted to get as close as we could?

Reply:
Yes, this is very important. E.g., JS doesn’t support 64-bit ints so any int8s have to be returned as strings. I’m sure there are similar problems in other languages with other drivers. Best to take the type at face value where possible. Plus switching to a wider type in Cockroach is only a metadata change so easy for the user to do later.

Reply:
@benesch I'm not sure exactly what your proposal is here. Are you arguing to keep the types as appearing as-close-as-possible to mysql or to make the visible type match the underlying type?